### PR TITLE
   Added some Objective C tests and minor bug fixes.

### PR DESCRIPTION
**Pull request by rmstar**
Objective-C tests: metadata, compression, keepalives, channel args.
Stress tests: network flap while streaming call in progress.

Bug fixes:
Stream gzip handling in interop server.
Keep alive, backoff time truncation bug in Obj-C layer.

### CODE REVIEW

**Comment by rmstar** [comment]
Known failures: #18416, #18968, #18967

**Comment by rmstar** [comment]
Known issue: #18416

File: src/objective-c/tests/UnitTests/APIv2Tests.m [hunk 1]
```diff
@@ -770,4 +843,319 @@ - (void)testFlowControlReadNonBlockingFailure {
...
+  GRPCCall2Internal *internalCall2 = call2->_firstInterceptor;
+  XCTAssertEqual(
+      internalCall1->_call->_wrappedCall->_pooledChannel->_wrappedChannel->_unmanagedChannel,
+      internalCall2->_call->_wrappedCall->_pooledChannel->_wrappedChannel->_unmanagedChannel);
+  [call1 finish];
+  [call2 finish];
+  [self waitForExpectations:rpcDone timeout:kTestTimeout];
+}
+
+- (void)testChannelReuseDifferentPath {
```
Comment (muxi) [inline src/objective-c/tests/UnitTests/APIv2Tests.m]: I didn't quite understand what this test is trying to test. Could you clarify (maybe with some comment on the top of this test)?

Comment (muxi) [reply src/objective-c/tests/UnitTests/APIv2Tests.m]: It looks like you are trying to test calls with different paths use the same channel, correct? But the paths are currently the same.

File: src/objective-c/tests/UnitTests/APIv2Tests.m [hunk 2]
```diff
@@ -770,4 +843,319 @@ - (void)testFlowControlReadNonBlockingFailure {
...
+  GRPCCall2Internal *internalCall2 = call2->_firstInterceptor;
+  XCTAssertEqual(
+      internalCall1->_call->_wrappedCall->_pooledChannel->_wrappedChannel->_unmanagedChannel,
+      internalCall2->_call->_wrappedCall->_pooledChannel->_wrappedChannel->_unmanagedChannel);
+  [call1 finish];
+  [call2 finish];
+  [self waitForExpectations:rpcDone timeout:kTestTimeout];
+}
+
+- (void)testChannelReuseDifferentCallSafety {
```
Comment (muxi) [inline src/objective-c/tests/UnitTests/APIv2Tests.m]: Consider merging this and the previous test. You can change path and safety at the same time, and assert that the two channels are equal

File: test/cpp/interop/interop_server.cc [hunk 3]
```diff
@@ -118,7 +118,8 @@ bool CheckExpectedCompression(const ServerContext& context,
               "Expected compression but got uncompressed request from client.");
       return false;
     }
-    if (!(inspector.GetMessageFlags() & GRPC_WRITE_INTERNAL_COMPRESS)) {
+    if (!(inspector.GetMessageFlags() & GRPC_WRITE_INTERNAL_COMPRESS) &&
+        received_compression != GRPC_COMPRESS_STREAM_GZIP) {
```
Comment (muxi) [inline test/cpp/interop/interop_server.cc]: Why this additional condition?

File: src/objective-c/tests/UnitTests/APIv2Tests.m (hunk 1 above)
Comment (rmstar) [reply src/objective-c/tests/UnitTests/APIv2Tests.m]: Oh sorry, the path's were the same so this test was useless. I've removed this testcase.

File: src/objective-c/tests/UnitTests/APIv2Tests.m (hunk 2 above)
Comment (rmstar) [reply src/objective-c/tests/UnitTests/APIv2Tests.m]: Removed the previous test case.

File: src/objective-c/tests/UnitTests/APIv2Tests.m [hunk 4]
```diff
@@ -770,4 +843,319 @@ - (void)testFlowControlReadNonBlockingFailure {
   [self waitForExpectationsWithTimeout:kTestTimeout handler:nil];
 }
 
+- (void)testChannelReuseIdentical {
```
[the remark this answers is not in the document]
Comment (rmstar) [reply src/objective-c/tests/UnitTests/APIv2Tests.m]: `ChannelTests` are different - they test the relationship between a GRPCPooledChannel and its wrapped channel. 
While I was looking at GRPCChannelPool, I found and fixed a bug in `wrappedCallWithPath:completionQueue:callOptions:` : we were ignoring the GRPCCompletionQueue param and using the default completion queue (singleton).

File: src/objective-c/tests/MacTests/StressTests.m [hunk 5]
```diff
@@ -519,21 +493,19 @@ - (void)testTimeoutOnFullDuplexCallWithV2API {
...
+    [request.responseParametersArray addObject:parameters];
+    request.payload.body = [NSMutableData dataWithLength:100];
 
-      [call writeMessage:request];
-      [call writeMessage:request];
-      [call finish];
-    });
+    [call writeMessage:request];
+    [call writeMessage:request];
+    [call finish];
```
Comment (muxi) [inline src/objective-c/tests/MacTests/StressTests.m]: Do we need sleep here too?

File: src/objective-c/tests/UnitTests/APIv2Tests.m (hunk 2 above)
Comment (muxi) [reply src/objective-c/tests/UnitTests/APIv2Tests.m]: That is the first half of the change, right? The second half is to change the `path` as well in this test, and verify that they still reuse the same channel.

File: src/objective-c/tests/UnitTests/APIv2Tests.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (muxi) [reply src/objective-c/tests/UnitTests/APIv2Tests.m]: Ok

File: src/objective-c/GRPCClient/GRPCCallOptions.h [hunk 6]
```diff
@@ -319,7 +319,8 @@ typedef NS_ENUM(NSUInteger, GRPCTransportType) {
 // HTTP/2 keep-alive feature. The parameter \a keepaliveInterval specifies the interval between two
 // PING frames. The parameter \a keepaliveTimeout specifies the length of the period for which the
 // call should wait for PING ACK. If PING ACK is not received after this period, the call fails.
-// Negative values are not allowed.
+// Negative values are invalid; setting these parameters to negative value will reset the
```
[the remark this answers is not in the document]
Comment (muxi) [reply src/objective-c/GRPCClient/GRPCCallOptions.h]: Ah yes. I was looking at a single commit and missed that.

**Pull request merged**